### PR TITLE
add support for bezier ENVs

### PR DESCRIPTION
**Pull request by friedolino78**
for me it works now.

### CODE REVIEW

**Comment by fundamental** [comment]
Sweet :D

File: src/osc-bridge/schema/test.json [hunk 1]
```diff
@@ -634,8 +634,8 @@
         "tooltip"  : "Attack Time",
         "scale"    : "logarithmic",
         "type"     : "f",
-        "range"    : [0.0f,41.0f],
-        "default"  : "0.0f"
+        "range"    : [0.0001f,41.0f],
```
Comment (fundamental) [inline src/osc-bridge/schema/test.json:637]: the github highlighting makes me think this should be `[0.0001,41.0]` i.e. without the 'f'.

File: build_config.rb [hunk 2]
```diff
@@ -83,6 +83,9 @@
   # Use Object class extension
   conf.gem :core => "mruby-object-ext"
 
+  # Use Enum class extension
+  conf.gem :core => "mruby-enum-ext"
```
Comment (fundamental) [inline build_config.rb:87]: Is this needed?

**Comment by fundamental** [comment]
I've got to still run the code to see how the control point manipulation feels, though there's a few questions posted to this and the zyn/zyn side.

File: src/mruby-zest/qml/Envelope.qml [hunk 3]
```diff
@@ -146,23 +157,28 @@ Widget {
...
 
             bound_points(env.xpoints,  0.0, 40950.0)
             bound_points(env.ypoints, -1.0, 1.0)
+            bound_points(env.cpoints, -2.0, 2.0)
 
             send_points() if mouse_enable
             update_nonfree_x(env.xpoints) if !mouse_enable
             update_nonfree_y(env.ypoints) if !mouse_enable
+            valueRef[2].value = env.cpoints if !mouse_enable
+            inspect env.cpoints
```
Comment (fundamental) [inline src/mruby-zest/qml/Envelope.qml:181]: line 181 Doesn't run...

**Comment by fundamental** [comment]
Hm, the beziers don't work as I would expect (also the area under the curve isn't rendered correctly, but that's fairly minor).

I expected something closer to inkscape's bezier curve functionality, though I guess the current results make sense given the code I read on the backend.

For something to be a bezier I'd say you would expect points to have 0, 1, or 2 control points and when there's two control points they should be symetrical (unless it's the sustain point?). Those control points would be set based on time and value, rather than just the value. So, I personally think that if this additional complexity is introduced there needs to be a way of upgrading points, having left/right control nodes, some symmetry to them, etc.

I'm still not 100% sure how intuitive of a feature this would be for sound design, though I get at least the base level of utility.

**Comment by friedolino78** [comment]
those are one quadratic bezier curve for each linear segment except the first one.
The control point is constrained to have its x coordinate in the middle between the endpoints. therefor there is only one additional parameter for each segment.
My main concern was to keep it backward compatible to old presets. If the control values are zero, everything is as linear as before.

**Comment by JohannesLorenz** [comment]
As a side note, I wonder how suited bezier splines would be for wave generation in OscilGen. There would be a lot of freedom to generate curves, but maybe they would all sound similar?

You could even modulate such waves by telling the control points to move into some specified direction.

**Comment by fundamental** [comment]
Assuming the off-tracker conversation was accurate, this PR is superseeded by the envNext PR series.

**Comment by friedolino78** [comment]
no, this is the ENVnext PR ;-)

**Pull request closed without merging**